extensions/amp-ad/0.1/amp-ad-batch-manager.js

```diff
@@ -0,0 +1,160 @@
+  * Note a side effect: the first ad with a given data-url to be processed will have its isBatchMaster_ variable set to true
+  * @param {Element} e The element which is to be batched
+  * @returns {AmpAdBatchManager}
+  */
+export function getBatchManager(e) {
+  // If this is our first imagead, create a map of responses for each value of 'data-url'
+  // This allows ads from multiple ad servers on the same page
+  if (!(e.win.hasOwnProperty('imageadBatchManagers'))) {
+    // Create an array of batch mangers. There will be one for each URL.
+    e.win.imageadBatchManagers = {};
+  }
+  if (!(e.url_ in e.win.imageadBatchManagers)) {
+    // This is the first time we've seen this URL, so this will be the master for this batch
+    e.isBatchMaster_ = true;
+    e.win.imageadBatchManagers[e.url_] = new AmpAdBatchManager(e);
+  } else {
+    e.isBatchMaster_ = false;
+  }
+  return e.win.imageadBatchManagers[e.url_];
+}
+
+export class AmpAdBatchManager {
+
+  /** @param {!Element} element */
+  constructor(element) {
+
+    /** @private {string} The base URL of the ad server */
+    this.url_ = element.url_;
+
+    /** @private {string} The full URL of the ad server, including any slot ids */
+    this.fullurl_ = this.url_;
+
+    /** @private {string} The number of timer ticks remaining before we give up waiting for an ad server response */
+    this.timeout_ = AD_BATCH_TIMEOUT;
+
+    /** @public {Object} responseData Map of JSON responses indexed by slot */
+    this.responseData = null;
+
+    // Scan the document for all elements with the same URL, and put them into the list to be batched
+    // @todo - the selector could be made a variable in order to generalise this to things other than imageads.
+    const elements = document.querySelectorAll('amp-ad[type=imagead]');
+    const slots = [];
+    for (let i = 0; i < elements.length; i++) {
+      const e = elements[i];
+
+      // If a slot is defined, add it to the list if it isn't there already
+      let s = e.getAttribute('data-slot');
+      if (s === null) {
+        s = 0;
+      }
+      let done = false;
+      for (const x in slots) {
+        if (slots[x] == s) {
+          done = true;
+        }
+      }
+      if (!done) {
+        slots.push(s);
+      }
+    }
+    // If any slots were defined, add them to the URL
+    if (slots.length) {
+      this.fullurl_ += '?s=' + slots.join(',');
+    }
+  }
+
+  /**
+   * Call this to instruct the batch manager to do the layout for an element. Prior to doing the layout, the batch manager
+   * will fetch the data (if this is the batch master) or wait for it to be ready (if not the batch master).
+   * @param {Element} e The element for which the layout is to be done
+   * @param {function} callback The callback to be provided to do the actual layout. The element's "jsondata_" variable will
+   *    be filled its with the data from the ad server before this is called.
+   * @returns {Promise}
+   */
+  doLayout(e, callback) {
```

Comment:
You're absolutely right, of course. Sorry - this is the first time I've used promises and I'm only just getting my head around them! The unnecessary callback is now gone.